include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+     * \{
+     */
+    using difference_type   = ptrdiff_t;                //!< Defaults to ptrdiff_t.
+    using value_type        = char_t;                   //!< The char type of the stream.
+    using reference         = char_t;                   //!< The char type of the stream.
+    using pointer           = void;                     //!< Has no pointer type.
+    using iterator_category = std::output_iterator_tag; //!< Pure output iterator.
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    fast_ostreambuf_iterator() noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator(fast_ostreambuf_iterator const &) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator(fast_ostreambuf_iterator &&) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator & operator=(fast_ostreambuf_iterator const &) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator & operator=(fast_ostreambuf_iterator &&) noexcept = default; //!< Defaulted.
+    ~fast_ostreambuf_iterator() noexcept = default; //!< Defaulted.
+
+    //!\brief Construct from a stream buffer.
+    explicit fast_ostreambuf_iterator(std::basic_streambuf<char_t, traits_t> & ibuf) :
+        stream_buf{reinterpret_cast<stream_buffer_exposer<char_t, traits_t> *>(&ibuf)}
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+            stream_buf->overflow(); // ensures that put area has space available
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator++()
+    {
+        return *this;
+    }
+    //!\overload
+    fast_ostreambuf_iterator & operator++(int)
+    {
+        return *this;
+    }
+    //!\}
+
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator*()
+    {
+        return *this;
+    }
+
+    //!\brief Writes a character to the associated output stream.
+    fast_ostreambuf_iterator & operator=(char_t const c)
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+        {
+            if (stream_buf->sputc(c) == traits_t::eof()) // overflow() [virtual], then write character
+                throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+        }
+        else
+        {
+            *stream_buf->pptr() = c;
+            stream_buf->pbump(1); // advance pptr() in put area without any checks
+        }
+        return *this;
+    }
+
+    //!\brief Returns `true if this iterator has encountered the end-of-file condition on output, `false` otherwise.
+    bool failed() const noexcept
+    {
+        return stream_buf->overflow() == traits_t::eof();
+    }
+
+    /*!\brief Writes a range to the associated output.
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     */
```

Comment:
Can you document the design decisions and the general mechanics of this function "for future generations"? :wink:

Reply:
[especially return value...]